packages/core/src/new-api/internal/module-builder.ts

```diff
@@ -101,6 +113,56 @@ export class IgnitionModuleBuilderImplementation<
     return future;
   }
 
+  public library<LibraryNameT extends string>(
+    libraryName: LibraryNameT,
+    options: LibraryOptions = {}
+  ): NamedLibraryDeploymentFuture<LibraryNameT> {
+    const id = options.id ?? libraryName;
+    const futureId = `${this._module.id}:${id}`;
+
+    this._assertUniqueContractId(futureId);
```

Comment:
Lets use a library specific version here, then clean up the duplication in a later PR.

Reply:
On the contrary, future ids are what are unique, it doesn't matter if they come from different type of futures.

For example, any two of those being present should lead to an error:

```ts
const c = m.contract("C");
const c2 = m.library("C");
const c3 = m.library("C3", {id: "C"});
```

Maybe John wanted to make the suggestion more specific?

I think I would go with an error message like 

`Duplicated id ${id} found in module ${moduleId}. Please fix this by passing an explicit id to the second future that's using it.`

Reply:
The id check is generic, every future id is compared with every other future id, whatever the type of future.

But the check needs parameterised in to variables:
1. The error message displayed `Your library has an id that has already been used` or some such
2. the function the check is being called from to allow filtering on the stack trace

As @alcuadrado mentions the message can be made generic, at that point you could just pass the filtering function, so maybe:

```
this._assertUniqueId(futureId, this.contract)
```